CIP-0XXX/README.md

```diff
@@ -0,0 +1,551 @@
+25  kesSignature = bstr
+26  operationalCertificate = bstr
+27  blockNumber = word32
+28  
+29  message = [
+30    messageId,
+31    messageBody,
+32    blockNumber,
+33    kesSignature,
+34    operationalCertificate
+35  ]
+30
+```
+
+### Client and server implementation
+
+This mini-protocol is designed with two goals in mind:
+* diffuse messages with high efficiency
+* protect from asymmetric resource attacks from the message consumer against the message provider
+
+The mini-protocol is based on two pull-based operations:
+- the message consumer asks for message ids,
+- and uses these ids to request a batch of messages (which it has not received yet)
+
+The client is responsible for initiating the mini-protocol with a peer node, but the server (i.e. the other node) is the one who asks for information.
+
+The client maintains a limited size FIFO queue of outstanding messages for each of the servers it is connected to, so does the server with a mirror FIFO queue of message ids:
+- the server asks for the next message ids and acknowledges for the previous message ids received (and removed from its queue).
+- the client removes the acknowledged ids from the FIFO queue it maintains for the server.
+- the server can download the content of the messages by giving an unordered list of ids to the client.
+
+The protocol supports blocking and non-blocking requests:
+- the client must reply immediately to a non-blocking request.
+- the server must wait until the client has at least one message available.
+- if the current queue of the server is empty, it must use a blocking request and a non-blocking request otherwise.
+
+### Protocol authentication
+
+#### Message authentication mechanism
+The message body is signed with the KES key of the SPO. This signature and the operational certificate of the SPO are appended to the message which is diffused.
+
+Before being diffused to other peers, an incoming message must be verified by the receiving node. This is done with the following steps:
+- Verify that the operational certificate is valid by checking that the KES verification key is signed by cold secret key.
+- Verify the KES signature of the message body with the KES verification key from the operational certificate.
+- Compute the SPO pool id by hashing the cold verification key from the operational certificate.
+- Make sure that this pool id is part of the stake distribution (the network layer will need to have access to this information).
+
+If any of these step fails, the message is considered as invalid, which is a protocol violation.
+
+> [!WARNING]  
+> We also probably need to make sure that the KES key used to sign is from the latest rotation:
+> - either the last seen opcert number in the block headers of the chain.
+> - or the last seen opcert number from a previous message diffused.
+> 
+> If the opcert number received is strictly lower than the previous one which has been seen, it should be considered as a protocol violation.
```

Comment:
I would also add a note saying that there might be needed an extra step of handshake to negotiate a mithril magic to forward the connection to the mithril node.

Reply:
Maybe it's a bit too much of an implementation detail?

Reply:
A standalone Mithril node will use its own `handshake`; it can introduce its own protocol parameters, but quite likely it will start with `NodeToNodeVersionData`:

```hs
data NodeToNodeVersionData = NodeToNodeVersionData
  { networkMagic  :: !NetworkMagic
  , diffusionMode :: !DiffusionMode
  , peerSharing   :: !PeerSharing
  , query         :: !Bool
  }
  deriving (Show, Typeable, Eq)
```

* `networkMagic` - I think that's useful for debugging, and making sure the `Mithril` node runs on the right network; but it should be different than the existing `networkMagic`s
* `diffusionMode` - this is useful if there are initiator-only nodes, e.g. a Mithril node running next to an edge node (a wallet)
* `peerSharing` - this will be useful, since we want peer sharing in the Mithril network
* `query` - that's useful for tools like `cardano-cli ping`, so I'd keep it.